### ISSUE TITLE
Authentication mechanism 

### ISSUE THREAD

**Issue by bbigras** [thecoshman/http#79]
Some kind of authentication could be nice.

My use case for this software is to allow people to get or upload a file on my LAN or with the internet. I would run the program just for a couple of minutes.

In my case it would be cool to have http generate a random token (maybe with expiration if people run http 24/24).

**Comment by thecoshman** [comment]
I mean sure... I don't see why we can't include this. It'll probably be one of these things that takes a while to work out what it should look like, but is fairly quick implement. 

From the sounds of it, you are just suggesting a simple 'token' be used, ie: a password that must be provided in order be allowed to upload files. 

I think this should extend to read access too, and be configurable to the folder/file level (ie, I can create a token that let's you read from `/foo`, another that grants read/write to `/bar`. Would also want a way to flag folders as public. And I guess these should nest, so that `/foo` can be protected, but `/foo/bar` be made public.

One big issue I see here though, is that it mostly seems pointless if we don't have encryption sorted. If your security model is "just enable the token for a short time, then disable it", you don't really have security. I don't want to be seen to be supporting/spreading further bad security practices.

In terms of configuring this, it's going to get rather hard to handle the complexity via cli args, so maybe just a single arg to a file (that doesn't have to be in the folder you are hosting) that has a definition for the permissions. What format that looks like... doesn't really matter...

**Comment by nabijaczleweli** [comment]
@bbigras please check the current `develop` branch with `--auth` or `--gen-auth` and see if it works for ye

**Comment by bbigras** [comment]
`--gen-auth` seems to work fine.

`--auth` is fine when used with both a username and a password but not when only used with a username.

```
$ cargo run --bin httplz -- --auth bob
[...]
Hosting "." on port 8000 without TLS and basic authentication using "bob" as username and no password...
Ctrl-C to stop.

[2019-01-24 17:04:35] 127.0.0.1:41534 requested to GET http://localhost:8000/ without authorisation
[2019-01-24 17:04:36] 127.0.0.1:41534 requested to GET http://localhost:8000/ with invalid credentials "bob:"
```

One thing that could be really cool with both options is to generate an URL with the credentials so I can copy/paste it over IM. I could do a PR for that. I would guess the hard part would be to enumerate all the local IP address.

**Comment by nabijaczleweli** [comment]
Hmm, looks like your browser treats empty password field as, well, empty password, as opposed to no password. That's an oversight, I'll merge the passwordless and clearpass cases.

Yeah, local IP enumeration is something that's been considered many times but it's just so ridiculously error-prone and usually-not-what-you-want that it's never made it in :v
Would writing out the credentials in the `username[:password]` form work for manageability?

**Comment by nabijaczleweli** [comment]
(@bbigras check new `develop`)

**Comment by nabijaczleweli** [comment]
(@bbigras, your input, if you'd be so kind?)

**Comment by bbigras** [comment]
I'm sorry I totally forgot.

`cargo run --bin httplz -- --auth bob` works now. :+1: 

> local IP enumeration is something that's been considered many times but it's just so ridiculously error-prone and usually-not-what-you-want that it's never made it in :v

yeah I can see that. now with docker and applications like virtualbox creating "fake" network interfaces.

Thanks for your patience.

**Comment by nabijaczleweli** [comment]
That's good to hear, now all that's left is for the old man to approve the PR :Р

Yeah, this and Even More F̬͚̮̕͟u҉̞̱̣͓̝͡n̝̮̥̮̱͈͖͟ ̼̳͖S͉͎͓̭̦͇t͕̰̙̖͖̟̮͇͟ų̗̦̖͕͈̪͝͞ͅf̶̻̰̺͚͡f̰͙̪̗͓̲̖͠ await any adventurer who attempts this once more.

dw fam, we've all been there

**Comment by nabijaczleweli** [comment]
Released in [`v1.5.0`](https://github.com/thecoshman/http/releases/tag/v1.5.0).
